File: src/main/java/jukebot/commands/Save.java
package jukebot.commands;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import jukebot.JukeBot;
import jukebot.audioutilities.AudioHandler;
import jukebot.utils.Command;
import jukebot.utils.CommandProperties;
import jukebot.utils.Context;
import net.dv8tion.jda.core.EmbedBuilder;

@CommandProperties(description = "DMs you the currently playing track. Specify `all` to save the queue", category = CommandProperties.category.MEDIA)
public class Save implements Command {

    public void execute(final Context context) {

        final AudioHandler player = context.getAudioPlayer();
        final AudioTrack currentTrack = player.player.getPlayingTrack();

        if (!player.isPlaying()) {
            context.sendEmbed("Not Playing", "Nothing is currently playing.");
            return;
        }

        if ("all".equalsIgnoreCase(context.getArgString())) {

            if (player.getQueue().isEmpty()) {
                context.sendEmbed("Queue is empty", "There are no tracks to save.");
                return;
            }

            StringBuilder sb = new StringBuilder();

            for (AudioTrack track : player.getQueue())
                sb.append(track.getInfo().title)
                        .append(" - ")
                        .append(track.getInfo().uri)
                        .append("\r\n");

            context.getAuthor().openPrivateChannel().queue(dm ->
                    dm.sendFile(
                            sb.toString().getBytes(), "queue.txt", null
                    ).queue(null, error ->
                            context.sendEmbed("Unable to DM", "Ensure your DMs are enabled.")
                    )
            );
        } else {
            context.getAuthor().openPrivateChannel().queue(dm ->
                    dm.sendMessage(
                            new EmbedBuilder()
                                    .setColor(JukeBot.embedColour)
                                    .setTitle(currentTrack.getInfo().title, currentTrack.getInfo().uri)
                                    .build()
                    ).queue(null, error ->
                            context.sendEmbed("Unable to DM", "Ensure your DMs are enabled.")
                    )
            );
        }

    }
}

File: src/main/java/jukebot/commands/Save.kt
package jukebot.commands

import jukebot.Database
import jukebot.entities.CustomPlaylist
import jukebot.framework.Command
import jukebot.framework.CommandCategory
import jukebot.framework.CommandProperties
import jukebot.framework.Context

@CommandProperties(description = "Adds the current track to a custom playlist", category = CommandCategory.QUEUE)
class Save : Command(ExecutionType.STANDARD) {

    override fun execute(context: Context) {
        val player = context.getAudioPlayer()
        val currentTrack = player.player.playingTrack

        if (!player.isPlaying) {
            return context.embed("Not Playing", "Nothing is currently playing.")
        }

        if (context.args.isEmpty()) {
            return context.embed("Save", "You need to provide the name of the playlist to add the track to.")
        }

        val playlist = Database.getPlaylist(context.author.idLong, context.argString)
            ?: return context.embed("Save", "That playlist doesn't exist.")

        if (playlist.tracks.size >= CustomPlaylist.TRACK_LIMIT) {
            return context.embed("Save", "You've hit the maximum amount of tracks for this playlist! (${CustomPlaylist.TRACK_LIMIT})")
        }

        playlist.tracks.add(currentTrack.makeClone())
        playlist.save()

        context.embed("Save", "Saved `${currentTrack.info.title}` to playlist `${playlist.title}`.")
    }
}
